src/fsm.cpp

```diff
@@ -171,3 +179,108 @@ void fsm::FSM::validate_final_states() const {
     }
 }
 
+bool fsm::FSM::evaluate(const char* input) {
+    fsm::String word(input);
+
+    for(int i = 0, l = word.size(); i < l; i++){
+        FSM::transition(word[i] - '0'); // convert char to int
+    }
+
+    bool flag = FSM::is_in_final_state();
+    FSM::restart();
+
+    return flag;
+}
+
+fsm::FSM fsm::FSM::operator!() const {
+    std::vector<fsm::State> newFinalStates;
+    fsm::FSM complementMachine = *this;
+
+    for(int i = 0, sz = get_states_count(); i < sz; i++){
+        if(std::find(final_states_.begin(),final_states_.end(),states_[i]) == final_states_.end()){
+            newFinalStates.push_back(states_[i]);
+        }
+    }
+
+    complementMachine.set_final_states(newFinalStates);
+
+    return complementMachine;
+}
+
+fsm::FSM fsm::FSM::operator&(const fsm::FSM &rhs) const {
+    fsm::FSM intersectionMachine;  // Initiate empty machine.
+    fsm::State comboState = get_current_state() + rhs.get_current_state();  // Get the initial state.
+    for(int i = 0, sz = get_alphabet_count(); i < sz; i++) { intersectionMachine.add_symbol(i); }  // Add the alphabet.
+    intersectionMachine.add_state(comboState);
+
+    fsm::fill(*this, rhs, intersectionMachine, comboState);
+
+    intersectionMachine.set_initial_state(comboState);
+    intersectionMachine.restart();
+
+    return intersectionMachine;
+}
+
+void fsm::fill(fsm::FSM m1, fsm::FSM m2, fsm::FSM &m3, fsm::State prevState) {
+    for(int i = 0, sz = m1.get_alphabet_count(); i < sz; i++){
+        auto m1old = m1, m2old = m2;
+        auto m3States = m3.get_states();
+
+        m1.transition(i);
+        m2.transition(i);
+
+        fsm::State comboState = m1.get_current_state() + m2.get_current_state();
+
+        if(std::find(m3States.begin(), m3States.end(), comboState) == m3States.end()){
+            m3.add_state(comboState);
+            if(m1.is_in_final_state() || m2.is_in_final_state()) { m3.add_final_state(comboState); }
```

Comment:
Shouldn't the final states of the intersection machine be the cartesian product of the final states of the two machines? That is: F_A∩B = F_A × F_B

So, I think that you should use `&&` not `||` here.